### PR TITLE
Chronos: train tcn model on gpu and speed up inference on cpu

### PR DESCRIPTION
**Pull request by smurf-1119**
## Description
uncompleted
train `tcn` model on gpu and speed up inference on cpu

https://qp-bigdl.readthedocs.io/en/nano-gpu/doc/Chronos/QuickStart/index.html

### CODE REVIEW

**Comment by TheaperDeng** [comment]
1. It should be better if we could use
  
  - `TSTrainer.compile` or
    
  - [BigDL-Nano Pytorch TorchNano Quickstart &mdash; BigDL documentation](https://bigdl.readthedocs.io/en/latest/doc/Nano/QuickStart/pytorch_nano.html) (if option1 is not valid)
    

to wrap TCN model

2. Merge the scritps from train and predict to 1 script where
  
  1). create a model
    
  2). train the model on gpu
    
  3). trace the model(`Trainer.trace`)
    
  4). test the latency (Use `model(x)` and `Evaluator.get_latency`where x is one sample data)
    
  
  Some reference link:
  
  [[Chronos-workshop/chronos_forecasting.ipynb at main · intel-analytics/Chronos-workshop · GitHub](https://github.com/intel-analytics/Chronos-workshop/blob/main/notebook/chronos_forecasting.ipynb)](https://github.com/intel-analytics/Chronos-workshop/blob/main/notebook/chronos_forecasting.ipynb)
  
  https://bigdl.readthedocs.io/en/latest/doc/PythonAPI/Chronos/evaluator.html#bigdl.chronos.metric.forecast_metrics.Evaluator.get_latency

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+from pytorch_lightning.callbacks import ModelCheckpoint
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:1]: `ModelCheckpoint` is not used

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import pytorch_lightning as pl
+from model import LitTCN
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:3]: `LitTCN` is not used

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import time
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:4]: not used

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import torch
+import matplotlib.pyplot as plt
+from bigdl.nano.pytorch import Trainer
+from bigdl.chronos.model.tcn import model_creator
+from bigdl.chronos.metric.forecast_metrics import Evaluator
+from tcn_config import *
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:10]: `tcn_config` is deleted

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            .roll(lookback=lookback,horizon=horizon)
+
+    tsdata_traindataloader = tsdata_train.to_torch_data_loader(batch_size=batch_size)
+    tsdata_valdataloader = tsdata_val.to_torch_data_loader(batch_size=batch_size, shuffle=False)
+    tsdata_testdataloader = tsdata_test.to_torch_data_loader(batch_size=batch_size, shuffle=False)
+
+    return tsdata_traindataloader,\
+           tsdata_valdataloader,\
+           tsdata_testdataloader,\
+           tsdata_test
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:53]: `tsdata_test` is never used

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+past_seq_len = lookback
+feature_seq_len = 1 # number of feature to predict
+optimizer = 'Adam'
+lr = 0.001
+batch_size = 32
+num_epochs = 3
+seed = 0
+device = "gpu"
+save_dir = './model'
+save_model_file = f'horizon:{horizon}lookback:{lookback}lr:{lr}_epoch:{num_epochs}' + 'tcn.pth'
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:28]: just remove these parameters

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    config = {'input_feature_num':input_feature_num,
+              'output_feature_num':output_feature_num,
+              'past_seq_len':past_seq_len,
+              'future_seq_len':feature_seq_len,
+              'kernel_size':3,
+              'repo_initialization':True,
+              'dropout':0.1,
+              'seed': seed,
+              'num_channels':[30]*7
+              }
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:76]: hardcode the numbers here directly.

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    tsdata_testdataloader = tsdata_test.to_torch_data_loader(batch_size=batch_size, shuffle=False)
+
+    return tsdata_traindataloader,\
+           tsdata_valdataloader,\
+           tsdata_testdataloader,\
+           tsdata_test 
+
+def test_time(model, dataloader):
+    for x, _ in dataloader:
+        res = model(x[0].unsqueeze(0))
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:57]:
```python
def predict_wraper(model, input_sample):
    model(input_sample)
```

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    trainer = pl.Trainer(max_epochs=3, val_check_interval=1.0,
+              accelerator='gpu', 
+              devices=1,
+              )
+    trainer.fit(lit_model, tsdata_traindataloader, tsdata_testdataloader)
+    
+    
+    print("original pytorch runtime (ms):", Evaluator.get_latency(test_time,lit_model, tsdata_testdataloader))
+    
+    speed_model = Trainer.trace(lit_model, accelerator="onnxruntime", input_sample=tsdata_testdataloader)
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:90]: You should use `tsdata_traindataloader` for `input_sample`

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    print("onnxruntime pytorch runtime (ms):", Evaluator.get_latency(test_time, speed_model, tsdata_testdataloader))
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:91]: onnxruntime latency (ms)

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    optimizer = torch.optim.Adam(lr=lr, params=model.parameters())
+    lit_model = Trainer.compile(model, loss, optimizer)
+    trainer = pl.Trainer(max_epochs=3, val_check_interval=1.0,
+              accelerator='gpu', 
+              devices=1,
+              )
+    trainer.fit(lit_model, tsdata_traindataloader, tsdata_testdataloader)
+    
+    
+    print("original pytorch runtime (ms):", Evaluator.get_latency(test_time,lit_model, tsdata_testdataloader))
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:88]: original pytorch latency (ms)

File: python/chronos/example/nano_qp/tcn/train_test.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    
+    x = None
+    for x, _ in tsdata_traindataloader:
+        break
+    input_sample = x[0].unsqueeze(0)
+    
+    print("original pytorch latency (ms):", Evaluator.get_latency(predict_wraper, lit_model, input_sample))
+    
+    speed_model = Trainer.trace(lit_model, accelerator="onnxruntime", input_sample=input_sample)
+    print("onnxruntime latency (ms):", Evaluator.get_latency(predict_wraper, speed_model, input_sample))
```
Comment (TheaperDeng) [inline python/chronos/example/nano_qp/tcn/train_test.py:72]: add a new line here

File: python/chronos/example/nano_qp/tcn/train_test.py (hunk 2 above)
Comment (TheaperDeng) [reply python/chronos/example/nano_qp/tcn/train_test.py:1]: pls fix this

File: python/chronos/example/nano_qp/tcn/train_test.py (hunk 5 above)
Comment (TheaperDeng) [reply python/chronos/example/nano_qp/tcn/train_test.py:33]: pls fix this
